### ISSUE TITLE
allow service workers to created nested dedicated workers

### ISSUE THREAD

**Issue by wanderview** [w3c/ServiceWorker#1529]
I think we've discussed this before, but I cannot find the issue.  It would be nice to allow service workers to created nested dedicated workers via `new Worker()`.  This would allow the service worker to perform CPU intensive operations, like custom decoding, without potentially blocking the thread being used to process further FetchEvents.

https://github.com/chromeos/static-site-scaffold-modules/issues/40#issuecomment-673591331 is an example of where this would be useful.

**Comment by jakearchibald** [comment]
Yeah, I think workers are currently linked to a root document, so that needs to be changed to "document or service worker global", or something common between the two.

**Comment by wanderview** [comment]
I think dedicated workers can be nested under shared workers right now?

**Comment by mfalken** [comment]
Would the idea be that the dedicated worker lives as long as the service worker is running?

**Comment by wanderview** [comment]
Correct.  Its lifetime would be bound by the lifetime of its parent, just like when creating a dedicated worker in a window or iframe.

**Comment by tophf** [comment]
FWIW, It would be useful for the new ManifestV3 browser extensions (currently implemented only in Chrome), which are built on service workers now so they've completely lost the ability to create workers in the background script.

**Comment by cohenerickson** [comment]
I would love to see this become implemented into the service worker spec, I think it would be beneficial for highly computational fetch requests.

**Comment by ian97531** [comment]
I created a similar issue over in WHATWG: https://github.com/whatwg/html/issues/8362

**Comment by js-choi** [comment]
GitHub comment

A while back in whatwg/html#8362, I made several comments describing real-world use cases for service workers spawning dedicated workers. I’m concisely summarizing them here too, since I this space might be more likely to trigger implementer interest.

An [MDN webpage on offline progressive web apps](https://developer.mozilla.org/en-US/docs/Web/Progressive_web_apps/Tutorials/js13kGames/Offline_Service_workers) says, “If you have heavy calculations to do, you can offload them from the main thread and do them in the worker, and receive results as soon as they are available.” In actuality, whenever a service worker needs to do CPU-expensive work, it must resort to complex coordination with at least one dedicated worker belonging to its origin’s browser tab(s), e.g., using Web Locks and some [leader-election](https://en.wikipedia.org/wiki/Leader_election) scheme. (This can get especially complicated when the leader tab gets closed, destroying its dedicated worker, and the newly elected leader tab must restore that dedicated worker’s state.)

This complexity affects the following use cases:

* **Offline SQLite, OPFS sync access, local storage**: Cache, IndexedDB, and cookies are currently the only persistent offline data storage available to service workers. A progressive web app that wishes to be available offline is unable to use without the aforementioned complex coordination with dedicated document workers; it must use IndexedDB or the like.
    * OPFS sync access is essential for WASM SQLite; it is currently impossible for an offline website’s service worker to fetch data from SQLite data in OPFS without coordinating with at least one dedicated worker belonging to its origin’s browser tab(s). See rhashimoto/wa-sqlite#95, rhashimoto/wa-sqlite#81, DallasHoff/sqlocal#39.
    * If service workers could use OPFS, webpages that are loaded from URLs typed into the address bar can be dynamically generated offline by the service workers. For example, a cloud document editor, wiki, CMS-based blog, or Pokédex web app that wished to make itself available offline could dump all of its documents’, wiki’s, or CMS’s data into a SQLite database, then use its service worker to dynamically generate the same web pages offline.
    * Furthermore, browser tabs would be able to use the same HTTP API both to read/write data in cloud storage and to read/write data to SQLite files or other documents in OPFS. When the service worker acts as a “transparent” mediator syncing cloud storage and with local caching, document scripts do not have to be concerned with whether documents are stored offline only, online only, or both.

* **Transparent decoding/decompression of unsupported file formats**: Service workers can intercept data fetches and transparently encode/decode data into new formats when the formats are unsupported by current browsers.
    * However, as said in https://github.com/whatwg/html/issues/8362#issuecomment-1271704657, custom decoding/decompression algorithms are CPU intensive and should be put in dedicated workers to avoid blocking the service workers. This is currently impossible.
    * A real-world case of this is [libjxl’s WASM decoder](https://github.com/libjxl/libjxl/tree/v0.11.x/tools/wasm_demo) for progressive enhancement of JPEG-XL images on a website. Ideally, the service worker would be able to perform the decoding process in a worker, to prevent blocking the service worker from other tasks (a tab→SW→worker→SW→tab dataflow path). However, because the service worker cannot currently spawn workers, it must use a [convoluted tab→SW→**tab→worker→tab**→SW→tab dataflow path](https://github.com/libjxl/libjxl/blob/v0.11.x/tools/wasm_demo/README.md).

* **Expensive RegExp processing for navigation routing**: A core use case of service workers is offline URL routing, but this routing may be expensive. As @wanderview mentioned in this issue’s original post, a real-world case of this is chromeos.dev’s >1.0-second synchronous microtask delay while processing complex regular expressions; see https://github.com/chromeos/static-site-scaffold-modules/issues/40#issuecomment-673591331.

**Comment by yoshisatoyanagisawa** [comment]
I am concerned about the complexity of detecting script changes in service workers when dynamic imports and dedicated workers are involved. Currently, service workers have a mechanism to automatically detect and reload when their scripts are updated. However, I believe that the introduction of dynamic imports and dedicated workers may pose a challenge to this mechanism. How can service workers reliably detect changes in scripts when they use dynamic imports or include dedicated workers?

**Comment by asutherland** [comment]
I think the most practical approach is to require dedicated workers spawned by ServiceWorkers to be controlled by the ServiceWorker that spawned them, including using the routes for that registration.  The ServiceWorker would be responsible for caching the worker scripts, ideally as part of the install phase, but it doesn't seem worth it to try and add enforcement mechanisms for that like we have for ServiceWorker scripts or for the update check mechanism to change.

The spec would also need to be clear that the controlled worker clients should not extend the underlying lifetime of the ServiceWorker similar to how ServiceWorker.postMessage from a ServiceWorker to another ServiceWorker should not extend the lifetime of the recipient ServiceWorker beyond the lifetime of the sender.

**Comment by tophf** [comment]
What about letting SW connect to an already running SharedWorker's port if it's trivial to implement? It'll be kinda weird though that `new SharedWorker()` may fail when the worker isn't running, so maybe a new API should be added e.g. to find the running worker via `self.clients` or a new one like `self.clientWorkers`...

**Comment by asutherland** [comment]
The clients API already exposes [and identifies shared worker clients](https://w3c.github.io/ServiceWorker/#dom-client-type).

**Comment by tophf** [comment]
> The clients API already exposes [and identifies shared worker clients](https://w3c.github.io/ServiceWorker/#dom-client-type).

Yes, but AFAICT it's unusable and `client.postMessage` doesn't reach SharedWorker because it only has a `connect` event and not `message`.

**Comment by js-choi** [comment]
<details>

<summary> I’d like to emphasize that service-workers-spawning-workers are important, regardless of whether shared are well supported.</summary>

This issue covers many use cases that shared workers do not address, particularly with those described in https://github.com/w3c/ServiceWorker/issues/1529#issuecomment-2641345477 – i.e., fetch handling that requires relatively long-running sync calls. Shared workers wouldn’t help with fetches using data from SQLite or other OPFS sync access; they wouldn’t help with transparent decoding/decompression of unsupported file formats or with async expensive RegExp processing for URL routing either.

If Chrome Android supported shared workers, that would help many other use cases that also currently require complex tab leader election and locking. But shared workers don’t address many important use cases in this thread. Even with broad shared-workers support—if we want to enable important fetch-related use cases, then we still need dedicated workers spawned by dedicated workers. So I think shared workers are largely off-topic from this issue’s topic.

</details>

***

@asutherland: I largely agree with the ideas in https://github.com/w3c/ServiceWorker/issues/1529#issuecomment-2643492051. I’m a little confused by what you mean by “enforcement mechanism”—do you mean how the last update check time can get updated, e.g., by `importScripts(urls)`? (Related comment by @jeffposnick from the dynamic-`import()` issue: https://github.com/w3c/ServiceWorker/issues/1585#issuecomment-825827457.) 

I can see it being a developer footgun if a change in a JS resource didn’t trigger an update in a service worker that used it as a dedicated worker…but it also probably would be a footgun if JS modules didn’t use the same module cache as the service worker’s statically `import`ed modules. 

I definitely think that `new Worker`’s SW-update-triggering behavior (or lack thereof) should be consistent whether or not it uses `type: "module"`. And I agree that it would be simpler and more consistent for module dedicated workers to just use the same module cache system as statically `import`ed modules. 

And that would mean that changes to dedicated-worker resources wouldn’t contribute to the service worker’s update check, whether or not the dedicated worker is a module. Each time that developers change one of their dedicated workers (or dynamic `import`s of those workers), they would just have to change a comment in the service worker or something similar.
